### DISCUSSION TITLE
Use JSON attributes instead of multiple data attributes in the DOM tree

### DISCUSSION THREAD

**Discussion by benedikt-buchert** [elbwalker/walkerOS#204] [Ideas]
Hi,

as I mentioned briefly on Measurecamp. It might be interesting to encode data-elb attributes as json payloads. These should be very easy to parse and can allow you to keep all tracking related attributes for each component in one element.

Instead of doing:

```
<div data-elb='promotion' data-elbaction='visible:view'>
      <h1 data-elb-promotion='name:Setting up tracking easily'>
        Setting up tracking easily
      </h1>
      <p data-elb-promotion='category:analytics'>Analytics</p>
 </div>
```

```
    <div data-elb="{ 'entity': 'promotion','action': {'visible': 'view'},'name': 'Setting up tracking easily','category': 'analytics'}">
      <h1>
        Setting up tracking easily
      </h1>
      <p>Analytics</p>
    </div>
```

Why would this be beneficial:
 - JSON is something developers know.
 - All tracking attributes for one component in once place.
 - JSON objects are easy to parse.
 - JSON objects support strings numbers and boolean by default. https://docs.elbwalker.com/tagging/properties#type-casting
 -  Multiple attributes also do not need to have a specific syntax: https://docs.elbwalker.com/tagging/properties#multiple-attributes
 - JSON supports arrays. https://docs.elbwalker.com/tagging/properties#arrays
 - Maybe it simplifies the library. (I'm not so deep into the code base.)

Why not do this:
- Don't change an established syntax.
- The current syntax helps you to place things in the right context.

**Comment by alexanderkirtzel** [comment]
Interesting approach. I thought you were talking about using JSON for data attributes only. Not for the entity and action itself, too:
```
<div data-elb="promotion" data-elb-promotion="{ 'name': 'Setting up tracking easily','category': 'analytics'}" />
```

The `init` may become tricky, when the walker can't just look for `[data-elbaction]` any longer.
I also know setups where the `[data-elb]` selector plays a huge role. But one can also use both.

What do you think: Is it an improvement to support data attributes in JSON format just for the properties? Or would it blow up the code _too much_ when devs could translate props to the current format?

**Reply by benedikt-buchert** [reply]
I think just having the attributes in a json format is already a huge simplification. from a dev perspective.
